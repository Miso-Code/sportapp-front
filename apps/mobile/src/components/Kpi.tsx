import React, { ComponentProps } from 'react'
import { StyleSheet, View } from 'react-native'
import { useTheme, Text, Icon } from 'react-native-paper'

import { ProgressChart, LineChart } from 'react-native-chart-kit'
import { hexToRgb } from '@/components/TimerRing'

interface BaseKpiProps extends ComponentProps<typeof View> {
	color?: string
}

interface ProgressKpiProps extends BaseKpiProps {
	type: 'progress'
	value: number
	max: number
	label: string
	valueSuffix?: string
	icon?: ComponentProps<typeof Icon>['source']
}

interface LineChartKpiProps extends BaseKpiProps {
	type: 'lineChart'
	data: number[]
	labels: string[]
	label: string
}

type KpiProps = ProgressKpiProps | LineChartKpiProps

export default function Kpi(props: KpiProps) {
	const theme = useTheme()
	const colorRGB = hexToRgb(props.color ?? theme.colors.primary)
	const textColorRGB = hexToRgb(theme.colors.onBackground)

	return (
		<View style={styles.container} {...props}>
			{props.type === 'progress' && (
				<>
					<ProgressChart
						data={{
							data: [
								Math.max(
									0,
									props.value < props.max
										? props.value
										: props.max
								) / props.max
							]
						}}
						width={200}
						height={200}
						strokeWidth={16}
						radius={75}
						chartConfig={{
							backgroundGradientFromOpacity: 0,
							backgroundGradientToOpacity: 0,
							color: (opacity = 1) =>
								`rgba(${colorRGB.r}, ${colorRGB.g}, ${colorRGB.b}, ${opacity})`
						}}
						hideLegend={true}
						data-testid='progressBar'
					/>
					<View style={styles.content}>
						{props.icon && (
							<Icon
								source={props.icon}
								size={35}
								color={props.color ?? theme.colors.primary}
								testID='icon'
							/>
						)}
						<Text variant='titleMedium'>
							{props.value} {props.valueSuffix}
						</Text>
						<Text variant='headlineSmall'>{props.label}</Text>
					</View>
				</>
			)}
			{props.type === 'lineChart' && (
				<>
					<LineChart
						data={{
							labels: props.labels,
							datasets: [
								{
									data: props.data
								}
							]
						}}
						width={350}
						height={200}
						chartConfig={{
							backgroundGradientFromOpacity: 0,
							backgroundGradientToOpacity: 0,
							color: (opacity = 1) =>
								`rgba(${colorRGB.r}, ${colorRGB.g}, ${colorRGB.b}, ${opacity})`,
							labelColor: (opacity = 1) =>
								`rgba(${textColorRGB.r}, ${textColorRGB.g}, ${textColorRGB.b}, ${opacity})`
						}}
						bezier
						data-testid='lineChart'
					/>
					<View style={styles.content}>
						<Text variant='headlineSmall'>{props.label}</Text>
					</View>
				</>
			)}
		</View>
	)
}

const styles = StyleSheet.create({
	container: {
		justifyContent: 'center',
		alignItems: 'center'
	},
	content: {
		position: 'absolute',
		alignItems: 'center'
	}
})
