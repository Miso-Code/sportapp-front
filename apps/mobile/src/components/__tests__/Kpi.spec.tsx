import React from 'react'
import renderer, { ReactTestRenderer } from 'react-test-renderer'

import Kpi from '../Kpi'

describe('Kpi', () => {
	let component: ReactTestRenderer

	beforeEach(() => {
		component = renderer.create(
			<Kpi type='progress' value={1} max={2} label='label' />
		)
	})

	afterEach(() => {
		component.unmount()
	})

	it('should render correctly', () => {
		expect(component.toJSON()).toMatchSnapshot()
	})

	it('should render an icon if provided', () => {
		component.update(
			<Kpi type='progress' value={1} max={2} label='label' icon='fire' />
		)
		expect(component.root.findByProps({ testID: 'icon' })).toBeTruthy()
	})

	it('should render a progress bar if type is progress', () => {
		expect(
			component.root.findByProps({ 'data-testid': 'progressBar' })
		).toBeTruthy()
	})

	it('should render a line chart if type is lineChart', () => {
		component.update(
			<Kpi
				type='lineChart'
				data={[1, 2, 3]}
				labels={['1', '2', '3']}
				label='test'
			/>
		)
		expect(
			component.root.findByProps({ 'data-testid': 'lineChart' })
		).toBeTruthy()
	})
})
