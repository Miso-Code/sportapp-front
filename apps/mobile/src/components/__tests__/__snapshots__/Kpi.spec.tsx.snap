// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Kpi should render correctly 1`] = `
<View
  label="label"
  max={2}
  style={
    {
      "alignItems": "center",
      "justifyContent": "center",
    }
  }
  type="progress"
  value={1}
>
  <View
    style={
      {
        "height": 200,
        "padding": 0,
        "width": 200,
      }
    }
  >
    <RNSVGSvgView
      bbHeight={200}
      bbWidth={200}
      focusable={false}
      height={200}
      style={
        [
          {
            "backgroundColor": "transparent",
            "borderWidth": 0,
          },
          {
            "flex": 0,
            "height": 200,
            "width": 200,
          },
        ]
      }
      width={200}
    >
      <RNSVGGroup
        fill={
          {
            "payload": 4278190080,
            "type": 0,
          }
        }
      >
        <RNSVGDefs>
          <RNSVGLinearGradient
            gradient={
              [
                0,
                0,
                1,
                0,
              ]
            }
            gradientTransform={null}
            gradientUnits={1}
            name="backgroundGradient"
            x1={0}
            x2={200}
            y1={200}
            y2={0}
          />
          <RNSVGLinearGradient
            gradient={
              [
                0,
                442978468,
                1,
                442978468,
              ]
            }
            gradientTransform={null}
            gradientUnits={1}
            name="fillShadowGradientFrom"
            x1={0}
            x2={0}
            y1={0}
            y2={200}
          />
        </RNSVGDefs>
        <RNSVGRect
          fill={
            {
              "brushRef": "backgroundGradient",
              "type": 1,
            }
          }
          height={200}
          propList={
            [
              "fill",
            ]
          }
          rx={0}
          ry={0}
          width="100%"
          x={0}
          y={0}
        />
        <RNSVGGroup
          fill={
            {
              "payload": 4278190080,
              "type": 0,
            }
          }
          matrix={
            [
              1,
              0,
              -0,
              1,
              100,
              100,
            ]
          }
        >
          <RNSVGGroup
            fill={
              {
                "payload": 4278190080,
                "type": 0,
              }
            }
          >
            <RNSVGPath
              d="M 0 -75 A 75 75 0 1 1 -0.471236 -74.99852 L -0.471236 -74.99852 A 75 75 0 1 0 0 -75 Z "
              fill={
                {
                  "payload": 4278190080,
                  "type": 0,
                }
              }
              propList={
                [
                  "stroke",
                  "strokeWidth",
                ]
              }
              stroke={
                {
                  "payload": 862408868,
                  "type": 0,
                }
              }
              strokeWidth={16}
            />
          </RNSVGGroup>
          <RNSVGGroup
            fill={
              {
                "payload": 4278190080,
                "type": 0,
              }
            }
          >
            <RNSVGPath
              d="M 0 -75 A 75 75 0 0 1 0 75 L 0 75 A 75 75 0 0 0 0 -75 Z "
              fill={
                {
                  "payload": 4278190080,
                  "type": 0,
                }
              }
              propList={
                [
                  "stroke",
                  "strokeWidth",
                  "strokeLinecap",
                  "strokeLinejoin",
                ]
              }
              stroke={
                {
                  "payload": 2154254500,
                  "type": 0,
                }
              }
              strokeLinecap={1}
              strokeLinejoin={1}
              strokeWidth={16}
            />
          </RNSVGGroup>
        </RNSVGGroup>
      </RNSVGGroup>
    </RNSVGSvgView>
  </View>
  <View
    style={
      {
        "alignItems": "center",
        "position": "absolute",
      }
    }
  >
    <Text
      style={
        [
          {
            "textAlign": "left",
          },
          {
            "color": "rgba(28, 27, 31, 1)",
            "writingDirection": "ltr",
          },
          [
            {
              "fontFamily": "System",
              "fontSize": 16,
              "fontWeight": "500",
              "letterSpacing": 0.15,
              "lineHeight": 24,
            },
            undefined,
          ],
        ]
      }
    >
      1
       
    </Text>
    <Text
      style={
        [
          {
            "textAlign": "left",
          },
          {
            "color": "rgba(28, 27, 31, 1)",
            "writingDirection": "ltr",
          },
          [
            {
              "fontFamily": "System",
              "fontSize": 24,
              "fontWeight": "400",
              "letterSpacing": 0,
              "lineHeight": 32,
            },
            undefined,
          ],
        ]
      }
    >
      label
    </Text>
  </View>
</View>
`;
